File: recursion/medium/24.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    
    def __repr__(self):
        return str(self.val)
    

def printList(node):
    if not node:
        return
    print(str(node.val) + ' ')
    printList(node.next)


class Solution:
    def swapPairs(self, head: ListNode):
        if not head or not head.next:
            return head

        new_head = head.next
        head.next = self.swapPairs(new_head.next)
        new_head.next = head
        return new_head


if __name__ == '__main__':
    s = Solution()

    # n4 = ListNode(4)
    # n3 = ListNode(3, n4)
    # n2 = ListNode(2, n3)
    # n1 = ListNode(1, n2)
    # printList(n1)
    # head = s.swapPairs(n1)
    # printList(head)

    
    n5 = ListNode(5)
    n4 = ListNode(4, n5)
    n3 = ListNode(3, n4)
    n2 = ListNode(2, n3)
    n1 = ListNode(1, n2)
    printList(n1)
    head = s.swapPairs(n1)
    printList(head)

File: linked_list/medium/24/24/24.cpp
#include <iostream>


struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
	ListNode(int x, ListNode *next) : val(x), next(next) {}
	void print() {
		ListNode* curr = this;
		while (curr != NULL) {
			std::cout << curr->val << "->";
			curr = curr->next;
		}
		std::cout << std::endl;
	}
};

class Solution {
public:
	ListNode* swapPairs(ListNode* head) {
		if (head == NULL || head->next == NULL) {
			return head;
		}
		ListNode* rest = swapPairs(head->next->next);
		ListNode* new_head = head->next;
		head->next->next = head;
		head->next = rest;
		return new_head;
	}
};
int main(){
	Solution s;
	ListNode* n0 = new ListNode(0);
	ListNode* n1 = new ListNode(1, n0);
	ListNode* n2 = new ListNode(2, n1);
	ListNode* n3 = new ListNode(3, n2);
	n3->print();
	s.swapPairs(n3)->print();




}

